src/core/particle_data.cpp

```diff
@@ -1258,310 +1258,48 @@ void auto_exclusions(int distance) {
-    type_array[Index.type[type]].id_list[i] = -1;
-
-  return ES_OK;
-}
-
-int reallocate_global_type_list(int size) {
-  if (size <= 0)
-    return ES_ERROR;
-  type_array = Utils::realloc(type_array, sizeof(TypeList) * size);
-  number_of_type_lists = size;
-  if (type_array == (TypeList *)0)
-    return ES_ERROR;
-
-  return ES_OK;
-}
-
-int find_particle_type(int type, int *id) {
-  int l_err = 1;
-  // type i not indexed, so no list for this particle exists
-  for (int i = 0; i < Type.max_entry; i++) {
-    if (Type.index[i] == type) {
-      l_err = 0;
-      break;
-    }
-  }
-  if (l_err) {
-    return ES_ERROR;
-  }
-  if (type_array[Index.type[type]].max_entry == 0) {
-    return ES_ERROR;
-  }
-  int rand_index = i_random(type_array[Index.type[type]].max_entry);
-  *id = type_array[Index.type[type]].id_list[rand_index];
-
-  return ES_OK;
-}
-
-int find_particle_type_id(int type, int *id, int *in_id) {
-  int l_err = 1;
-  // type i not indexed, so no list for this particle exists
-  for (int i = 0; i < Type.max_entry; i++) {
-    if (Type.index[i] == type) {
-      l_err = 0;
-      break;
-    }
-  }
-  if (l_err) {
-    return ES_ERROR;
-  }
-  if (type_array[Index.type[type]].max_entry == 0)
-    return ES_ERROR;
-
-  int rand_index = i_random(type_array[Index.type[type]].max_entry);
-  if (id == (int *)0 && in_id == (int *)0)
-    return ES_ERROR;
-  else {
-    *in_id = rand_index;
-    *id = type_array[Index.type[type]].id_list[*in_id];
-    return ES_OK;
-  }
-}
-
-int add_particle_to_list(int part_id, int type) {
-  int l_err = 1;
-  int already_in = 0;
-  //	int already_in_other_list = 0;
-  // type i not indexed, so no list for this particle exists
-  for (int i = 0; i < Type.max_entry; i++) {
-    if (Type.index[i] == type) {
-      l_err = 0;
-      break;
-    }
-  }
-  if (l_err) {
-    return NOT_INDEXED;
-  }
-
-  int in_type = Index.type[type];
-  int max = type_array[in_type].max_entry;
-  for (int i = 0; i < max; i++) {
-    if (type_array[in_type].id_list[i] == part_id) {
-      already_in = 1;
-      break;
-    }
-  }
-  if (already_in) {
-    return ES_OK;
-  }
-
-  if (max >= (double)type_array[in_type].cur_size / 2.0)
-    if (reallocate_type_array(type) == ES_ERROR)
-      return ES_ERROR;
-
-  // add particle id to list:
-  type_array[in_type].id_list[max] = part_id;
-  type_array[in_type].max_entry++;
-  return ES_OK;
-}
-
-int gc_status(int type) {
-  int l_err = 1;
-  // type i not indexed, so no list for this particle exists
-  for (int i = 0; i < Type.max_entry; i++) {
-    if (Type.index[i] == type) {
-      l_err = 0;
-      break;
+int get_random_p_id(int type){
+    if(ParticleTypeMap.at(type).size()==0)
+        throw std::runtime_error("No particles of given type could be found");
+    int returned_p_id=0;
+    int rand_index = i_random(ParticleTypeMap.at(type).size());\
+    int actual_i_th_p_id_in_map=0;
+    for( int p_id :ParticleTypeMap[type]){
```

Comment:
```c++
return *std::advance(ParticleTypeMap[type].begin(), rand_index);
```

Reply:
This is a way to draw a random element from an unordered set.
According to http://en.cppreference.com/w/cpp/iterator/advance std::advance needs a Bidirectional Iterator. However the unordered set only is guaranteed to have a forward iterator (http://www.cplusplus.com/reference/unordered_set/unordered_set/). Therefore I cannot use std::advance on the unordered set.

Reply:
No

Reply:
True, the problem was std::advance returns void and advances the iterator itself...
    std::unordered_set<int>::iterator it=particle_type_map[type].begin();
    std::advance(it,rand_index);
    return *it;
works